### ISSUE TITLE
Finetuning

### ISSUE THREAD

**Issue by rlenain** [jasonppy/VoiceCraft#70]
Hello, and thanks a lot for your great work.

I am finetuning your 830M model on custom data, but getting to a point where I am overfitting fairly quickly.

I was wondering, are there training details you recommend for finetuning? Maybe specific learning rate, or parts of the model to freeze? More generally, do you have an idea of how many hours would be required in order to not overfit with this size of model?

Thanks a lot!

**Comment by jasonppy** [comment]
Thanks! 

I haven't tried finetuning a lot. You could use the 330M model. I tried finetuning on as small as 550h libritts and it seems that it doesn't overfit - I used the AdamW optimizer (directly supported by the codebase), lr=1e-5, max_num_tokens=20000. I haven't tried freezing part of the model but that sounds very reasonable (LORA is another option)

**Comment by danablend** [comment]
> Thanks!
> 
> I haven't tried finetuning a lot. You could use the 330M model. I tried finetuning on as small as 550h libritts and it seems that it doesn't overfit - I used the AdamW optimizer (directly supported by the codebase), lr=1e-5, max_num_tokens=20000. I haven't tried freezing part of the model but that sounds very reasonable (LORA is another option)

Hey! Curious if you would happen to have an idea of which layers may be ideal to apply LoRA fine tuning to and if so how many of them?

For example, would it be suitable to train LoRA layers on the first N and last N attention layers in the TransformerEncoder based on trial/error (could be found with sweeping), or would a different approach be more ideal to capture a new speaker's characteristics such as tonality and prosody?

Thank you!

**Comment by rishikksh20** [comment]
Hi @jasonppy 
I am finetuning the 330M TTS model on multi-lingual data, and here is the tensorboard 
![image](https://github.com/jasonppy/VoiceCraft/assets/4656872/5e431612-2abd-410b-982d-d37bc614ffb3)
With finetuning on a single A6000 with max num of token 10K with accum grad 24, do you think curve looks good ?
And what is the good value for loss and top10 ?

**Comment by jasonppy** [comment]
Looks pretty good! I think it's worth generating a few sample.

I don't really know what is a good value for loss and top10, as different data have different levels of difficulty. For english only data, audiobooks usually gets cb1 top10 60 or higher on training, gigaspeech gets ~55 for cb1.

max_num_token is for 1 gpu, so if you run it on 4 gpus, the effective token consumed for one gradient update is 40k.

**Comment by rishikksh20** [comment]
@jasonppy Yes 52-55 % start producing good voice, I will also let you know it works great with multi-lingual data. I finetuned this on 3 lang data and even when I combine the phonemes from different languages into a single line it is able to speak perfectly like code-switch language models.
Kudos for the great work

**Comment by rikabi89** [comment]
> @jasonppy Yes 52-55 % start producing good voice, I will also let you know it works great with multi-lingual data. I finetuned this on 3 lang data and even when I combine the phonemes from different languages into a single line it is able to speak perfectly like code-switch language models. Kudos for the great work

Would mind sharing how you setup a custom dataset for fine-tuning please such any such script etc?

**Comment by rishikksh20** [comment]
Hi @jasonppy 
For multi-lingual, I don't do anything extra I just rely on Espeak-ng phonemes.  Create a dataset based on Phonemizer language-based phonemes, mix all multi-lingual datasets, and mix the phoneme vocabs using pre-processing scripts of this repo. And simply update the Phoneme embedding of the pre-train model and trained.

**Comment by rishikksh20** [comment]
But I think when we include lots of languages and accents it might not work as intended because many IPA phonemes are shared between the languages, so might be needed to pass any token along with transcripts to signify the Accent and language for better pronunciation.

**Comment by rikabi89** [comment]
> Hello, and thanks a lot for your great work.
> 
> I am finetuning your 830M model on custom data, but getting to a point where I am overfitting fairly quickly.
> 
> I was wondering, are there training details you recommend for finetuning? Maybe specific learning rate, or parts of the model to freeze? More generally, do you have an idea of how many hours would be required in order to not overfit with this size of model?
> 
> Thanks a lot!

Hi would mind sharing how you setup a custom dataset for fine-tuning please such any such script etc?

**Comment by rishikksh20** [comment]
@thivux Yes it's sensitive to the hyper-params but it gives good performance on certain parameters

**Comment by Me210400** [comment]
> Hi @jasonppy I am finetuning the 330M TTS model on multi-lingual data, and here is the tensorboard ![image](https://private-user-images.githubusercontent.com/4656872/322707928-5e431612-2abd-410b-982d-d37bc614ffb3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2ODkxMjQsIm5iZiI6MTcyMjY4ODgyNCwicGF0aCI6Ii80NjU2ODcyLzMyMjcwNzkyOC01ZTQzMTYxMi0yYWJkLTQxMGItOTgyZC1kMzdiYzYxNGZmYjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDNUMTI0MDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2RkMDE5YjQ5Mzg1YWYyMmQyYjMxZTY5OTJlNmUxODI0ZWMyNTc5MzE5ZGY0OGNkNDVlZjEwODhlYWRmZmVmZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.vL6fUdSuP2O9SjQukLeKIBqhqU0SC_QmKzFpQjgz31Y) With finetuning on a single A6000 with max num of token 10K with accum grad 24, do you think curve looks good ? And what is the good value for loss and top10 ?

how you prepare your custom dataset?

**Comment by Magauiya** [comment]
Hi @rishikksh20! Thank you for sharing insights for multilanguage scenario! Have you tried training from scratch and compare with finetuned one? Also how do you assess (metrics) the quality of your fine-tuned model?

**Comment by rishikksh20** [comment]
Nope I only finetune the model, training from scratch will required too much data and compute. 
I listened to the files generated from diverse set of paragraph and listen, is the only way to check the quality.
